torch/_decomp/decompositions.py

```diff
@@ -3312,66 +3359,66 @@ def _compute_source_index(scale, dst_index, align_corners):
 ) -> Tensor:
     # get dimensions of original image
-    _, n_channels, in_h, in_w = input.shape
-
-    # Calculate horizontal and vertical scaling factor
-    h_scale_factor = _compute_scale(in_h, output_size[0], align_corners, scales_h)
-    w_scale_factor = _compute_scale(in_w, output_size[1], align_corners, scales_w)
+    n_batch, n_channels = input.shape[:2]
+    inp_size = input.shape[2:]
+    d = len(inp_size)
 
     _, dtype = utils.elementwise_dtypes(
-        input, type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT
+        input,
+        type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
     )
-    # We have to create arange with int64 dtype and use .to in order to avoid
-    # additional kernels creation in inductor and get a perf slowdown
-    i = torch.arange(output_size[0], device=input.device).to(dtype=dtype)
-    j = torch.arange(output_size[1], device=input.device).to(dtype=dtype)
-
-    x_f32 = _compute_source_index(w_scale_factor, j, align_corners).clamp(min=0.0)
-    y_f32 = _compute_source_index(h_scale_factor, i, align_corners).clamp(min=0.0)
-    y_f32 = y_f32.unsqueeze(-1)
-
-    x = x_f32.to(torch.int64)
-    y = y_f32.to(torch.int64)
-
-    # We are using torch.where instead of torch.clamp below due to an expected failure
-    # in test_aot_autograd_symbolic_exhaustive_nn_functional_interpolate_bilinear_cpu_float32 test
-    # torch.ops.aten.clamp.default(add, None, sub) on int64 input tensor is returning float32 and
-    # fails with torch.ops.aten._unsafe_index.Tensor(primals_1, [None, None, _to_copy_1, clamp_2])
-    # RuntimeError: _unsafe_index found unexpected index type Float
-    # xp1 = (x + 1).clamp(max=in_w - 1); yp1 = (y + 1).clamp(max=in_h - 1)
-    xp1 = torch.where(x < in_w - 1, x + 1, x)
-    yp1 = torch.where(y < in_h - 1, y + 1, y)
-
-    v1 = aten._unsafe_index(input, [None, None, y, x])
-    v2 = aten._unsafe_index(input, [None, None, y, xp1])
-    v3 = aten._unsafe_index(input, [None, None, yp1, x])
-    v4 = aten._unsafe_index(input, [None, None, yp1, xp1])
-
-    dtype = torch.float32 if not input.is_floating_point() else input.dtype
-    if not input.is_floating_point():
-        v1 = v1.to(dtype)
-        v2 = v2.to(dtype)
-        v3 = v3.to(dtype)
-        v4 = v4.to(dtype)
 
-    yscale = (y_f32 - y).clamp(0.0, 1.0).to(dtype)
-    xscale = (x_f32 - x).clamp(0.0, 1.0).to(dtype)
+    def get_values(inp, out, scales, nsqueeze):
+        # First Calculate scaling factor
+        scale_factor = _compute_scale(inp, out, align_corners, scales)
+        # We have to create arange with int64 dtype and use .to in order to avoid
+        # additional kernels creation in inductor and get a perf slowdown
+        i = torch.arange(out, device=input.device).to(dtype=dtype)
+
+        x_f32 = _compute_source_index(scale_factor, i, align_corners).clamp(min=0.0)
+        x_f32 = x_f32.reshape(x_f32.shape[0], *[1] * (nsqueeze))
+        x = x_f32.to(torch.int64)
```

Comment:
Actually, the accuracy issues seem to come from the dtypes used for `x`. In the eager implementation we use cast `x` above to `acc_type`, which is `f64` for CPU.

Let's not fix this in this PR to unblock it. Let's fix it in a follow-up PR.

cc @vfdev-5

Reply:
After a discussion with Victor,  the current implementation matches eager, so nvm. The accuracy issues happen on onnx, so it's not super high-prio.

Reply:
As discussed elsewhere, on eager on CPU, casting rule is the following:
- https://github.com/pytorch/pytorch/blob/b8e71cf3022e701604ea1f0c381c0b9ccf8743be/aten/src/ATen/OpMathType.h#L14-L50
and here is where it is used for linear interp:
- https://github.com/pytorch/pytorch/blob/b8e71cf3022e701604ea1f0c381c0b9ccf8743be/aten/src/ATen/native/cpu/UpSampleKernel.cpp#L1112-L1137

So, I'd say no problems with `x_f32` dtype. I agree that `x_f32` is not a good name which actually means a floating value (not integral).